filebeat/docs/fields.asciidoc

```diff
@@ -44305,7 +44305,7 @@ Examples include Beats. Agents may also run on observers. ECS agent.* fields sha
 Extended build information for the agent.
 This field is intended to contain any build information that a data source may provide, no specific formatting is required.
 
-type: keyword
+type: wildcard
```

Comment:
So, I'm not entirely sure how we want to document these -- do we need to change the doc generator to specify that sometimes this is a wildcard (if you're Elastic licensed), sometimes it's a keyword (if you're OSS)? Or should we add some sort of top-level documentation that says "all wildcards are keywords in OSS"?

@andrewkroh @webmat @ebeahan any thoughts?

Reply:
Yeah we have an issue in ECS for 1.8 to document the "fallback" idea for OSS implementations, and what it means at a high level. We can probably cover the Beats and Logstash implementations there as well.

Perhaps Beats docs can link to that?

We can make sure this page is ready in advance of the Stack 7.11 release.

Reply:
Just to answer your question a little more directly

> do we need to change the doc generator...

I don't think we need to. The "OSS" fallback is not an official variant of ECS, it's a best effort workaround for purely OSS implementations. People could even use that to adopt ECS on 6.x if they have complex setups with many versions. But we're not making tons of efforts to fully support this. For example basic types that don't really have a fallback will simply not be supported.

This may deserve a mention somewhere in the Beats documentation, but I'm not even sure if there's a section about OSS Beats.

Reply:
👍 -- also, should we mention that we're using "experimental" features of ECS somewhere in our docs?

Reply:
Not sure. It's likely something that could flip flop from one release to another. For instance, "multi-user" stage 3 RFC is now merged, and will be part of mainline ECS in 1.8. Wildcard is approved, and it's a matter of time until it's merged as well.

So next release this notice may go away. And the following release, if we have another "stage 2" change we're pushing to adopt aggressively, the notice may need to come back... So I think it could add more confusion than anything.

It would be a fair warning to have though, so I wouldn't oppose if you want to add this.

Reply:
I do think this deserves a mention in the documentation so that users are aware that OSS beats will automatically be using `keyword` rather than `wildcard`. And that the same happens if you're using older, pre-wildcard, ES versions.

@dedemorton Do you have a recommendation on where to include such a notice?

Reply:
This is a tough call. It sounds like saying something in the field description (where it would be the most noticeable) is not a good idea because it would be repetitive (plus our docgen process limits us).

We can put a statement at the top-level page in the field reference for each Beat, but IMO it's likely to be overlooked by users. 

When are users most likely to care about the type difference? Will they be able to continue using the pre-built dashboards? Are there any upgrade implications if they decide to move to basic later on? Will there be a log message to let users know that keyword is being used instead of wildcard (and the implications/limitations)?

Reply:
@dedemorton `wildcard` is a drop-in replacement in terms of functionality. Everything will continue to work functionally the same (including dashboards). Only the performance characteristics are changing. So everything keeps working, some things faster, some things slower (check out the table in [the blog post for a breakdown](https://www.elastic.co/blog/find-strings-within-strings-faster-with-the-new-elasticsearch-wildcard-field)).

A user upgrading Basic Beats from 7.10 to 7.11 or from Beats OSS to Beats Basic 7.11+ will go through the same non-event. The `wildcard` and `keyword` types are compatible and mixed indices being queried together will work flawlessly. 

Users will only be able to fully take advantage of the performance improvements enabled by `wildcard` once the older data has aged out, though.